client/blocks/payment-request/use-checkout-subscriptions.js

```diff
@@ -0,0 +1,241 @@
+						? 'invalid_shipping_address'
+						: 'fail',
+					shippingOptions: [],
+				} );
+			}
+			clearPaymentRequestEventHandler( 'shippingAddressChange' );
+		};
+		const onShippingSelectedRate = ( forSuccess = true ) => () => {
+			const handlers = eventHandlers.current;
+			const shipping = currentShipping.current;
+			const billingData = currentBilling.current;
+			if (
+				handlers.shippingOptionChange &&
+				! shipping.isSelectingRate &&
+				isProcessing
+			) {
+				const updateObject = forSuccess
+					? {
+							status: 'success',
+							total: getTotalPaymentItem( billingData.cartTotal ),
+							displayItems: normalizeLineItems(
+								billingData.cartTotalItems
+							),
+					  }
+					: {
+							status: 'fail',
+					  };
+				handlers.shippingOptionChange.updateWith( updateObject );
+				clearPaymentRequestEventHandler( 'shippingOptionChange' );
+			}
+		};
+		const onProcessingPayment = () => {
+			const handlers = eventHandlers.current;
+			if ( handlers.sourceEvent && isProcessing ) {
+				const response = {
+					type: responseTypes.SUCCESS,
+					meta: {
+						billingData: getBillingData( handlers.sourceEvent ),
+						paymentMethodData: getPaymentMethodData(
+							handlers.sourceEvent,
+							currentPaymentRequestType.current
+						),
+						shippingData: getShippingData( handlers.sourceEvent ),
+					},
+				};
+				return response;
+			}
+			return { type: responseTypes.SUCCESS };
+		};
+		const onCheckoutComplete = ( checkoutResponse ) => {
+			const handlers = eventHandlers.current;
+			let response = { type: responseTypes.SUCCESS };
+			if ( handlers.sourceEvent && isProcessing ) {
+				const {
+					paymentStatus,
+					paymentDetails,
+				} = checkoutResponse.processingResponse;
+				if ( paymentStatus === responseTypes.SUCCESS ) {
+					completePayment( handlers.sourceEvent );
+				}
+				if (
+					paymentStatus === responseTypes.ERROR ||
+					paymentStatus === responseTypes.FAIL
+				) {
+					abortPayment( handlers.sourceEvent );
+					response = {
+						type: responseTypes.ERROR,
+						message: paymentDetails?.errorMessage,
+						messageContext: noticeContexts.EXPRESS_PAYMENTS,
+						retry: true,
+					};
+				}
+				clearPaymentRequestEventHandler( 'sourceEvent' );
+			}
+			return response;
+		};
+		if ( canMakePayment && isProcessing ) {
+			const unsubscribeShippingRateSuccess = onShippingRateSuccess(
+				onShippingRatesEvent
+			);
+			const unsubscribeShippingRateFail = onShippingRateFail(
+				onShippingRatesEventFail
+			);
+			const unsubscribeShippingRateSelectSuccess = onShippingRateSelectSuccess(
+				onShippingSelectedRate()
+			);
+			const unsubscribeShippingRateSelectFail = onShippingRateSelectFail(
+				onShippingRatesEventFail
+			);
+			const unsubscribePaymentProcessing = onPaymentProcessing(
+				onProcessingPayment
+			);
+			const unsubscribeCheckoutCompleteSuccess = onCheckoutAfterProcessingWithSuccess(
+				onCheckoutComplete
+			);
```

Comment:
I'm currently working on some fixes to the event handling (see [this issue](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/4048)) and while testing (the yet unpublished PR) I noticed that the current setup of the 3DS support for the payment_request integration is _not closing the payment request modal_ before pulling up the 3DS modal (it's happening outside of the branch I'm working on too). I don't know if you've noticed this in testing yet or not but it's pretty bad UX.

The good news is there's a solution:
```suggestion
			const unsubscribeCheckoutCompleteSuccess = onCheckoutAfterProcessingWithSuccess(
				onCheckoutComplete,
				5
			);
```

The checkout event handlers support priority (defaults to 10)  so adding a priority on this event will ensure it executes before the payment intents modal registered on the same event.

The bad news, is that you won't see the 3DS modal come up at all if you're testing against WooCommerce blocks because currently the first observer registered on the event returning a success response or error response aborts the event loop and no other events will execute (thus the payment intents event observer won't fire). However, I've verified this is fixed in the branch I'm working on to address [this issue](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/4048). 

Implications for merging/releasing this PR:

- As soon as I get a PR up for [this issue](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/4048), I'll let you folks know so you can test against it.
- It does mean you would have to target whatever release of blocks includes the above PR for full 3DS support.

Reply:
Here's the [related PR](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4158). Currently it's ready for testing (if you see it while it's still draft that's just because I haven't finished the description and testing notes for it yet).

Reply:
Here's how it works with the above fix applied (adding a priority) and against the branch in the linked Woo Blocks PR:

![CleanShot 2021-04-30 at 13 22 11](https://user-images.githubusercontent.com/1429108/116731111-49f78680-a9b7-11eb-9781-e76c81689078.gif)

Reply:
Here's an example of the bug where the payment request modal doesn't close (still testing against the Woo Blocks branch in the PR, but without the priority number fix):

![CleanShot 2021-04-30 at 13 29 32](https://user-images.githubusercontent.com/1429108/116731929-5af4c780-a9b8-11eb-9553-89f184b3245e.gif)

Reply:
Can you elaborate @ricardo? Given the first gif above which is the working example there already appears to be sufficient loading indicators. It's not clear to me where you are wanting additional loading indicators?

Reply:
Oh, I think I know what you are referring to now. After 3DS submission and while the checkout is processing you want some way of preventing shoppers from clicking the express payment method button again (and also have a loading indicator of sorts)?

Reply:
> Oh, I think I know what you are referring to now. After 3DS submission and while the checkout is processing you want some way of preventing shoppers from clicking the express payment method button again (and also have a loading indicator of sorts)?

Yes. Sorry, I didn't see your latest comment when I replied.

Reply:
`isProcessing` will be false while the 3DS modal is open. The point you are referring to is the actual checkout submission and redirect (payment methods have handed off at that point). This seems like something that would be useful to be handled by the checkout layer vs the payment method. Perhaps having some sort of loading mask over any express payment method buttons that prevent them from being interacted with while the redirect is happening.

If you scroll down in your test, you'll see that the checkout submit button has a loading indicator on it while the request is being processed (and until the redirect). Checkout has awareness of the stage of the flow so I think it would make sense for it to handle this. It'd also cover the scenario where there might be multiple express payment methods and we don't want _any_ interacted with while the processing is happening.

If you agree, would you mind creating an issue in the blocks repository for this? It might need some design input as well and I can get that.

Reply:
> `isProcessing` will be false while the 3DS modal is open. 

Yes, I was thinking about setting it to `true` only after the confirmation.

> Perhaps having some sort of loading mask over any express payment method buttons that prevent them from being interacted with while the redirect is happening.

This would be ideal! 👍 

> Checkout has awareness of the stage of the flow so I think it would make sense for it to handle this. It'd also cover the scenario where there might be multiple express payment methods and we don't want any interacted with while the processing is happening.

Yeah, I noticed that, and found weird that the express payment method didn't follow the same behavior. So I was wondering if there was a "native" way to do it. But since there isn't, I believe we should definitely have that in the checkout/cart layer.

Issue created: 4159-gh-woocommerce/woocommerce-gutenberg-products-block

Reply:
Thank you for bringing this to our attention @nerrad!

I just wanted to note that @ricardo changed the implementation in some significant ways in WCPay and this is no longer an issue there. I think the reason that fixed the issue is that more of the flow is delegated to our JS code as opposed to some of the Blocks APIs, but I'm not sure.
I also don't know if that has any detremental effects on the Blocks functionality in terms of other payment methods or managing the checkout flows — maybe @ricardo looked into that?

I'm working on replacing some of the Blocks implementation with our pre-existing AJAX APIs in #1545 and hopefully that will have similar results for this PR.

It's great that you're working on a fix though! I tried to find something simple to fix this, but couldn't figure anything out 😅